contributor-docs/code-change-guide.md

```diff
@@ -0,0 +1,519 @@
+
+In each java project or subproject, the `build.gradle` file starts with:
+
+```groovy
+
+apply plugin: 'org.apache.beam.module'
+
+applyJavaNature( ... )
+```
+
+Relevant usage of `BeamModulePlugin` includes:
+* Manage Java dependencies
+* Configure projects (Java, Python, Go, Proto, Docker, Grpc, Avro, an so on)
+  * Java -> `applyJavaNature`; Python -> `applyPythonNature`, and so on
+  * Define common custom tasks for each type of project
+    * `test`: run Java unit tests
+    * `spotlessApply`: format java code
+
+## Code paths
+
+The following are example code paths relevant for SDK development:
+
+* `sdks/java` Java SDK
+  * `sdks/java/core` Java core
+  * `sdks/java/harness` SDK harness (entrypoint of SDK container)
+
+* `runners` runner supports, written in Java. For example,
+  * `runners/direct-java` Java direct runner
+  * `runners/flink-java` Java Flink runner
+  * `runners/google-cloud-dataflow-java` Dataflow runner (job submission, translation, etc)
+    * `runners/google-cloud-dataflow-java/worker` Worker on Dataflow legacy runner
+
+* `sdks/python` contains the setup file and scripts to trigger test-suites
+  * `sdks/python/apache_beam` actual beam package
+    * `sdks/python/apache_beam/runners/worker` SDK worker harness entrypoint, state sampler
+    * `sdks/python/apache_beam/io` I/O connectors
+    * `sdks/python/apache_beam/transforms` most "core" components
+    * `sdks/python/apache_beam/ml` Beam ML
+    * `sdks/python/apache_beam/runners` runner implementations and wrappers
+    * ...
+
+* `sdks/go` Go SDK
+
+* `.github/workflow` GitHub action workflows (for example, tests run under PR). Most
+  workflows run a single Gradle command. Check which command is running for
+  a test so that you can run the same command locally during development.
+
+## Environment setup
+
+To set up local development environments, first see the [Contributing guide](../CONTRIBUTING.md) .
+If you plan to use Dataflow, see the [Google Cloud documentation](https://cloud.google.com/dataflow/docs/quickstarts/create-pipeline-java) to setup `gcloud` credentials.
+
+To check if your environment is set up, follow these steps:
+
+Your `PATH` needs to have the following elements configured.
+
+* A Java environment (any supported Java version, Java8 preferably as of 2024).
+  * This environment is needed for all development, because Beam is a Gradle project that uses JVM.
+  * Recommended: Use [sdkman](https://sdkman.io/install) to manage Java versions.
+* A Python environment (any supported Python version)
+  * Needed for Python SDK development
+  * Recommended: Use [`pyenv`](https://github.com/pyenv/pyenv) and
+    a [virtual environment](https://docs.python.org/3/library/venv.html) to manage Python versions.
+* A Go environment. Install the latest Go version.
+  * Needed for Go SDK development and SDK container change (for all SDKs), because
+  the container entrypoint scripts are written in Go.
+* A Docker environment.
```

Comment:
You don't need docker (and, consequently, go) for cross-language anymore.

Reply:
yeah, I noted "some cross-language functionality (if you run a SDK container image)". To clarify further, added note "not required since Beam 2.53.0")